File: packages/simulation-core/Sources/SimulationCore/ProSeasonArcRules.swift
import Foundation

/// 시즌을 한 문장으로 남기는 닫힌 제목 집합.
public enum ProSeasonArcTitle: String, Codable, Sendable {
    case firstHalfAce = "first_half_ace"
    case dominant = "dominant"
    case longTunnel = "long_tunnel"
    case lateRecovery = "late_recovery"
    case autumnDoorClosed = "autumn_door_closed"
    case autumnChampion = "autumn_champion"
    case autumnRunnerUp = "autumn_runner_up"
    case autumnEliminated = "autumn_eliminated"
    case autumnUnavailable = "autumn_unavailable"
    case quiet = "quiet"
}

public enum ProSeasonArcRules {
    public static func title(
        gameLines: [ProGameLine],
        postseason: ProPostseasonState?
    ) -> ProSeasonArcTitle {
        if let postseason {
            switch postseason.result {
            case .champion: return .autumnChampion
            case .runnerUp: return .autumnRunnerUp
            case .eliminated: return .autumnEliminated
            case .didNotQualify: return .autumnDoorClosed
            case .unavailable: return .autumnUnavailable
            case .inProgress: break
            }
        }

        let first = split(gameLines, in: 1...12)
        let second = split(gameLines, in: 13...24)
        let firstQuality = quality(first)
        let secondQuality = quality(second)
        switch (firstQuality, secondQuality) {
        case (.good, .poor): return .firstHalfAce
        case (.good, .good): return .dominant
        case (.poor, .poor): return .longTunnel
        case (.poor, .good): return .lateRecovery
        default: return .quiet
        }
    }

    public static func contentID(for title: ProSeasonArcTitle) -> String {
        "pro.arc.\(title.rawValue)"
    }

    private enum SplitQuality { case good, even, poor }

    private static func split(_ lines: [ProGameLine], in weeks: ClosedRange<Int>) -> (runs: Int, outs: Int) {
        lines.filter { weeks.contains($0.week) }.reduce((0, 0)) { partial, line in
            (partial.0 + line.runsAllowed, partial.1 + line.outs)
        }
    }

    private static func quality(_ split: (runs: Int, outs: Int)) -> SplitQuality {
        guard split.outs >= 27 else { return .even }
        let ra9 = split.runs * 27_000 / split.outs
        if ra9 < 3_500 { return .good }
        if ra9 >= 4_500 { return .poor }
        return .even
    }
}

File: apps/android/game-core/src/main/kotlin/com/solkim/baseball/core/pro/ProSeasonArcRules.kt
package com.solkim.baseball.core.pro

/** 시즌을 한 문장으로 남기는 닫힌 제목 집합. */
public enum class ProSeasonArcTitle(public val wire: String) {
    FIRST_HALF_ACE("first_half_ace"),
    DOMINANT("dominant"),
    LONG_TUNNEL("long_tunnel"),
    LATE_RECOVERY("late_recovery"),
    AUTUMN_DOOR_CLOSED("autumn_door_closed"),
    AUTUMN_CHAMPION("autumn_champion"),
    AUTUMN_RUNNER_UP("autumn_runner_up"),
    AUTUMN_ELIMINATED("autumn_eliminated"),
    AUTUMN_UNAVAILABLE("autumn_unavailable"),
    QUIET("quiet"),
}

public object ProSeasonArcRules {
    public fun title(gameLines: List<ProGameLine>, postseason: ProPostseasonState?): ProSeasonArcTitle {
        if (postseason != null) {
            when (postseason.result) {
                ProPostseasonResult.CHAMPION -> return ProSeasonArcTitle.AUTUMN_CHAMPION
                ProPostseasonResult.RUNNER_UP -> return ProSeasonArcTitle.AUTUMN_RUNNER_UP
                ProPostseasonResult.ELIMINATED -> return ProSeasonArcTitle.AUTUMN_ELIMINATED
                ProPostseasonResult.DID_NOT_QUALIFY -> return ProSeasonArcTitle.AUTUMN_DOOR_CLOSED
                ProPostseasonResult.UNAVAILABLE -> return ProSeasonArcTitle.AUTUMN_UNAVAILABLE
                ProPostseasonResult.IN_PROGRESS -> Unit
            }
        }
        val first = split(gameLines, 1..12)
        val second = split(gameLines, 13..24)
        return when (quality(first) to quality(second)) {
            SplitQuality.GOOD to SplitQuality.POOR -> ProSeasonArcTitle.FIRST_HALF_ACE
            SplitQuality.GOOD to SplitQuality.GOOD -> ProSeasonArcTitle.DOMINANT
            SplitQuality.POOR to SplitQuality.POOR -> ProSeasonArcTitle.LONG_TUNNEL
            SplitQuality.POOR to SplitQuality.GOOD -> ProSeasonArcTitle.LATE_RECOVERY
            else -> ProSeasonArcTitle.QUIET
        }
    }

    public fun contentId(title: ProSeasonArcTitle): String = "pro.arc.${title.wire}"

    private enum class SplitQuality { GOOD, EVEN, POOR }

    private fun split(lines: List<ProGameLine>, weeks: IntRange): Pair<Int, Int> =
        lines.filter { it.week in weeks }.fold(0 to 0) { partial, line ->
            (partial.first + line.runsAllowed) to (partial.second + line.outs)
        }

    private fun quality(split: Pair<Int, Int>): SplitQuality {
        val runs = split.first
        val outs = split.second
        if (outs < 27) return SplitQuality.EVEN
        val ra9 = runs * 27_000 / outs
        if (ra9 < 3_500) return SplitQuality.GOOD
        if (ra9 >= 4_500) return SplitQuality.POOR
        return SplitQuality.EVEN
    }
}
